inst/assets/shinyvalidate.js

```diff
@@ -61,43 +61,53 @@ const bindingStrategy = {
 strategies.push(bindingStrategy);
 
 /**
- * This strategy detects .shiny-input-container at or above the el, and uses
- * Bootstrap 3 classes to display validation messages.
+ * This strategy detects a .form-control within the input el, then:
+ * (1) toggles .is-invalid on the control (BS4+) or .has-error to the container (BS3) 
+ * (2) adds/removes .invalid-feedback (BS4+) or .help-block (BS3)
+ * 
+ * https://getbootstrap.com/docs/4.4/components/forms/#server-side
  */
-const bs3Strategy = {
-  findInputContainer: function(el) {
+const bootstrapStrategy = {
+  isBS3: function() {
+    return window.jQuery ? window.jQuery.fn.tab.Constructor.VERSION.match(/^3\./) : false;
```

Comment:
Rather than tying this to jQuery, we could use `$.fn.tooltip.Constructor.VERSION` to determine this (checking that `$.fn.tooltip` isn't falsy).